lib/reline/line_editor.rb

```diff
@@ -406,110 +313,239 @@ def multiline_off
+    @output.write prompt
+  end
+
+  def render_differential
+    unless @dialog_initialzed
+      update_dialogs
+    end
+
+    editor_cursor_x, editor_cursor_y = editor_cursor_position
+
+    rendered_lines = @rendered_screen_cache || []
+    new_lines = wrapped_lines.flatten[screen_scroll_top, screen_height].map do |l|
+      [[0, Reline::Unicode.calculate_width(l, true), l]]
+    end
+    if @menu_info
+      @menu_info.list.sort!.each do |item|
+        new_lines << [[0, Reline::Unicode.calculate_width(item), item]]
       end
-      @buffer_of_lines[@line_index] = @line
-      @rest_height = 0 if @scroll_partial_screen
-    else
-      line = modify_lines(whole_lines)[@line_index]
-      render_partial(prompt, prompt_width, line, 0)
-      if finished?
-        scroll_down(1)
-        Reline::IOGate.move_cursor_column(0)
-        Reline::IOGate.erase_after_cursor
+      @menu_info = nil # TODO: do not change state here
+    end
+
+    @dialogs.each_with_index do |dialog, index|
+      next unless dialog.contents
+
+      x_range, y_range = dialog_range dialog, editor_cursor_y - screen_scroll_top
+      y_range.each do |row|
+        next if row < 0 || row >= screen_height
+        dialog_rows = new_lines[row] ||= []
+        dialog_rows[index + 1] = [x_range.begin, dialog.width, dialog.contents[row - y_range.begin]]
+      end
+    end
+
+    if new_lines != rendered_lines
+      Reline::IOGate.hide_cursor
+      num_lines = [[new_lines.size, rendered_lines.size].max, screen_height].min
+      scroll_down(num_lines - 1 - @cursor_y) if (num_lines - 1 - @cursor_y) > 0
+      @cursor_y = num_lines - 1
+      num_lines.times do |i|
+        rendered_line = rendered_lines[i] || []
+        line_to_render = new_lines[i] || []
+        next if rendered_line == line_to_render
+
+        Reline::IOGate.move_cursor_down i - @cursor_y
+        @cursor_y = i
+        unless rendered_lines[i]
+          Reline::IOGate.move_cursor_column 0
+          Reline::IOGate.erase_after_cursor
+        end
+        render_line_differential(rendered_line, line_to_render)
       end
+      @rendered_screen_cache = new_lines
+      Reline::IOGate.show_cursor
     end
+    y = editor_cursor_y - screen_scroll_top
+    Reline::IOGate.move_cursor_column editor_cursor_x
+    Reline::IOGate.move_cursor_down y - @cursor_y
+    @cursor_y = y
+    new_lines.size - @cursor_y
+  end
+
+  def current_row
+    wrapped_lines.flatten[editor_cursor_y]
+  end
+
+  def upper_space_height
+    @cursor_y - 1
+  end
+
+  def rest_height
+    screen_height - @cursor_y - @cursor_base_y - 1
+  end
+
+  def handle_cleared
+    return unless @cleared
+
+    @cleared = false
+    @rendered_screen_cache = nil
+    Reline::IOGate.clear_screen
+    @screen_size = Reline::IOGate.get_screen_size
+    @cursor_base_y = 0
+    @cursor_y = 0
+    scroll_into_view
+    render_differential
+  end
+
+  def rerender
+    handle_cleared
```

Comment:
> It also looks like we could be calling render_differential twice under the current implementation?

Nice catch. I remove `render_differential` from `handle_cleared`. Finally when pasting is complete, rerender will be called with `@in_pasting == false` from `reline.rb`.

Reply:
I want to leave handle_cleared because I want to remove `handle_cleared` from `rerender` in the future.
But changing it now needs `test_key_actor_emacs.rb` that depends on instance variable `@cleared` to be changed.

Plan and reason
```ruby
private def ed_clear_screen(key)
  @cleared = true
end

def rerender
  handle_cleared
  render_differential unless @in_pasting
end
```
↓
```ruby
private def ed_clear_screen(key)
  @rendered_screen_cache = nil
  Reline::IOGate.clear_screen
  @screen_size = Reline::IOGate.get_screen_size
  @cursor_base_y = 0
  @cursor_y = 0
end
# We need handle_resize(`def resize` is implemented) and handle_interrupt(not implemented, causing deadlock and thread problem) to handle signals correctly
# We don't need handle_cleared. clear is not a signal
def rerender
  render_differential unless @in_pasting
end
```